lib/collection.js

```diff
@@ -0,0 +1,290 @@
+    },
+
+    /**
+     * @param {Object} conditions
+     * @returns {?Model}
+     */
+    findWhere: function (conditions) {
+        return this.find(this._isMatchConditions.bind(this, conditions));
+    },
+
+    /**
+     * @param {String} attr
+     * @returns {Array}
+     */
+    pluck: function (attr) {
+        return this.map(function (model) {
+            return model.get(attr);
+        });
+    },
+
+    /**
+     * @param {(Model|Object|Array.<(Model|Object)>)} models
+     * @param {Object} [options]
+     * @param {Number} [options.at]
+     * @returns {Collection}
+     */
+    add: function (models, options) {
+        var at = options && options.at;
+        models = Array.isArray(models) ? models : [models];
+        models.forEach(function (model, index) {
+            var currentOptions;
+            model = this._prepareModel(model);
+            if (!this._isExists(model)) {
+                if (at !== undefined) {
+                    currentOptions = clone(options);
+                    currentOptions.at = at + index;
+                }
+                this._addModel(model, currentOptions);
+            }
+        }, this);
+
+        return this;
+    },
+
+    /**
+     * @param {(Model|Array.<Model>)} models
+     * @returns {Collection}
+     */
+    remove: function (models) {
+        models = Array.isArray(models) ? models : [models];
+        models.forEach(function (model) {
+            if (this._isExists(model)) {
+                this._removeModel(model);
+            }
+        }, this);
+
+        return this;
+    },
+
+    /**
+     * @param {Array.<Model>} models
+     * @returns {Collection}
+     */
+    set: function (models) {
+        models = models.map(this._prepareModel, this);
+
+        while (this._models.length) {
+            this._removeModel(this.at(0));
+        }
+        models.forEach(this._addModel, this);
+
+        return this;
+    },
+
+    /**
+     * @param {Object} conditions
+     * @param {Model} model
+     * @returns {Boolean}
+     */
+    _isMatchConditions: function (conditions, model) {
+        return Object.keys(conditions).every(function (key) {
+            return model.attributes[key].isEqual(conditions[key]);
+        });
+    },
+
+    /**
+     * @param {(Model|Object)} model
+     * @returns {Model}
+     */
+    _prepareModel: function (model) {
+        return model instanceof this.modelType ? model : new this.modelType(model);
+    },
+
+    /**
+     * @param {Model} model
+     * @returns {Boolean}
+     */
+    _isExists: function (model) {
+        return model.getId() !== null && Boolean(this.get(model.getId()));
+    },
+
+    /**
+     * @param {Model} model
+     * @param {Object} [options]
+     */
+    _addModel: function (model, options) {
+        var at = options && options.at || this._models.length;
+        this._models.splice(at, 0, model);
+        if (model.getId() !== null) {
```

Comment:
what if id field will appear in model in runtime,  after model is inserted in collection?
id to model index won't be updated